In [7]:
import pandas as pd

In [8]:
p_and_l = pd.read_excel(r"C:\Users\bhrra\Desktop\nifty100\data\raw\profitandloss.xlsx",skiprows=1)
doc = pd.read_excel(r"C:\Users\bhrra\Desktop\nifty100\data\raw\documents.xlsx",skiprows=1)
p_and_c = pd.read_excel(r"C:\Users\bhrra\Desktop\nifty100\data\raw\prosandcons.xlsx",skiprows=1)

In [9]:
companies_df = pd.read_excel(r"C:\Users\bhrra\Desktop\nifty100\data\raw\companies.xlsx",skiprows=1)
bs_df = pd.read_excel(r"C:\Users\bhrra\Desktop\nifty100\data\raw\balancesheet.xlsx",skiprows=1)

In [10]:
print(bs_df.head(2))
print("_"*100)
print(companies_df.head(2))
print("_"*105)
print(bs_df.columns)
print("_"*105)
print(companies_df.columns)

    id company_id      year  equity_capital  reserves  borrowings  \
0  136        ABB  Dec 2012            21.0       626           0   
1  137        ABB  Mar 2014            21.0       767           0   

   other_liabilities  total_liabilities  fixed_assets  cwip  investments  \
0                260                907           109     1            0   
1                351               1139            98     1            0   

   other_asset  total_assets  
0          798           907  
1         1040          1139  
____________________________________________________________________________________________________
           id                                       company_logo  \
0         ABB   https://mkt.in/static/mkt-icons/nifty100/ABB.png   
1  ADANIENSOL  https://m.economictimes.com/thumb/msid-1173715...   

                 company_name  \
0            Abbott India Ltd   
1  Adani Energy Solutions Ltd   

                                          chart_link  \
0  https

In [11]:
print(bs_df.info())
print("_"*110)
print(companies_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1312 entries, 0 to 1311
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 1312 non-null   int64  
 1   company_id         1312 non-null   object 
 2   year               1312 non-null   object 
 3   equity_capital     1312 non-null   float64
 4   reserves           1312 non-null   int64  
 5   borrowings         1312 non-null   int64  
 6   other_liabilities  1312 non-null   int64  
 7   total_liabilities  1312 non-null   int64  
 8   fixed_assets       1312 non-null   int64  
 9   cwip               1312 non-null   int64  
 10  investments        1312 non-null   int64  
 11  other_asset        1312 non-null   int64  
 12  total_assets       1312 non-null   int64  
dtypes: float64(1), int64(10), object(2)
memory usage: 133.4+ KB
None
_______________________________________________________________________________________________________

In [12]:
print(companies_df.isnull().sum())
nulls = companies_df.isnull().sum()
nulls[nulls > 0]


id                 0
company_logo       1
company_name       0
chart_link         0
about_company      0
website            1
nse_profile        1
bse_profile        1
face_value         1
book_value         1
roce_percentage    1
roe_percentage     2
dtype: int64


company_logo       1
website            1
nse_profile        1
bse_profile        1
face_value         1
book_value         1
roce_percentage    1
roe_percentage     2
dtype: int64

In [13]:
companies_df[companies_df.isnull().any(axis=1)]

,id,company_logo,company_name,chart_link,about_company,website,nse_profile,bse_profile,face_value,book_value,roce_percentage,roe_percentage
46,INDIGO,https://rb.gy/rv6gr8\n,Interglobe Aviation Ltd,https://in.tradingview.com/chart/?symbol=NSE%3...,Interglobe Aviation Ltd (Indigo) is India's la...,http://www.goindigo.in/,https://www.nseindia.com/get-quotes/equity?sym...,https://www.bseindia.com/stock-share-price/int...,10.0,97.0,24.5,NaN
61,LTIM,NaN,LTIMindtree Ltd,https://mkt.in/static/mkt-icons/nifty100/ABB.png,LTIMindtree Limited is an Indian multinational...,https://www.ltimindtree.com/,https://www.nseindia.com/get-quotes/equity?sym...,https://www.bseindia.com/stock-share-price/lti...,1.0,711.0,34.1,26.04
91,TVSMOTOR,https://mkt.in/static/mkt-icons/nifty100/TVSMO...,TVS Motor Company Ltd,https://in.tradingview.com/chart/qGsydD2w/?sym...,'TVS Motor Company Ltd (TVSM) is engaged in ma...,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [14]:
print(p_and_l.head(5))
print('-'*100)
print(doc.head(5))
print("-"*100)
print(p_and_c.head(5))

   id company_id      year  sales  expenses  operating_profit  opm_percentage  \
0  61        ABB  Dec 2012   1653      1451             202.0            12.0   
1  62        ABB  Mar 2014   2276      2009             267.0            12.0   
2  63        ABB  Mar 2015   2289      1977             312.0            14.0   
3  64        ABB  Mar 2016   2614      2250             365.0            14.0   
4  65        ABB  Mar 2017   2903      2505             398.0            14.0   

   other_income  interest  depreciation  profit_before_tax  tax_percentage  \
0            33         0            19                215            33.0   
1            49         0            22                295            33.0   
2            48         0            15                344            34.0   
3            50         3            14                398            36.0   
4            57         2            16                436            37.0   

   net_profit    eps  dividend_payout  
0   

In [15]:
p_and_l.columns

Index(['id', 'company_id', 'year', 'sales', 'expenses', 'operating_profit',
       'opm_percentage', 'other_income', 'interest', 'depreciation',
       'profit_before_tax', 'tax_percentage', 'net_profit', 'eps',
       'dividend_payout'],
      dtype='object')

In [16]:
len(p_and_l["company_id"].unique())

100

In [17]:
len(p_and_c["company_id"].unique())

5

In [18]:
len(doc["company_id"].unique())

99

In [19]:
print(set(p_and_l["company_id"].unique()) - set(doc["company_id"].unique()).intersection(set(p_and_l["company_id"].unique())))

{'DIVISLAB'}


In [20]:
p_and_l[p_and_l["company_id"] == 'DIVISLAB']
# doc[doc["company_id"] == 'DIVISLAB']

,id,company_id,year,sales,expenses,operating_profit,opm_percentage,other_income,interest,depreciation,profit_before_tax,tax_percentage,net_profit,eps,dividend_payout
352,425,DIVISLAB,Mar 2013,2145,1329,816.0,38.0,45,3,77,781,23.0,602,23.0,33.0
353,426,DIVISLAB,Mar 2014,2532,1514,1018.0,40.0,68,3,92,991,22.0,773,29.0,34.0
354,427,DIVISLAB,Mar 2015,3115,1942,1173.0,38.0,38,3,136,1072,21.0,852,32.0,31.0
355,428,DIVISLAB,Mar 2016,3776,2354,1423.0,38.0,93,5,118,1393,19.0,1126,42.0,24.0
356,429,DIVISLAB,Mar 2017,4064,2615,1450.0,36.0,72,3,123,1395,24.0,1060,40.0,25.0
357,430,DIVISLAB,Mar 2018,3891,2622,1269.0,33.0,107,2,142,1231,29.0,877,33.0,30.0
358,431,DIVISLAB,Mar 2019,4946,3072,1874.0,38.0,154,5,169,1855,27.0,1353,51.0,31.0
359,432,DIVISLAB,Mar 2020,5394,3568,1826.0,34.0,186,7,186,1819,24.0,1377,52.0,31.0
360,433,DIVISLAB,Mar 2021,6969,4102,2867.0,41.0,57,2,256,2666,26.0,1984,75.0,27.0
361,434,DIVISLAB,Mar 2022,8960,5075,3885.0,43.0,112,2,312,3684,20.0,2960,112.0,27.0


In [21]:
a =(p_and_l["year"].unique())
print(a)

['Dec 2012' 'Mar 2014' 'Mar 2015' 'Mar 2016' 'Mar 2017' 'Mar 2018'
 'Mar 2019' 'Mar 2020' 'Mar 2021' 'Mar 2022' 'Mar 2023' 'Mar 2024' 'TTM'
 'Mar 2013' 'Dec 2013' 'Dec 2014' 'Dec 2015' 'Dec 2016' 'Dec 2017'
 'Dec 2018' 'Dec 2019' 'Dec 2020' 'Dec 2021' 'Mar 2023 15' 'Jun 2013'
 'Jun 2014' 'Jun 2015' 'Mar 2016 9m' 'Mar 2011' 'Mar 2012' 'Dec 2022'
 'Dec 2023' 'Sep 2011' 'Sep 2014' 'Sep 2015' 'Sep 2016' 'Sep 2017'
 'Sep 2018' 'Sep 2019' 'Sep 2020' 'Sep 2021' 'Sep 2022' 'Sep 2023'
 'Sep 2024']


In [22]:
p_and_l["company_id"].unique()

array(['ABB', 'ADANIENSOL', 'ADANIENT', 'ADANIGREEN', 'ADANIPORTS',
       'ADANIPOWER', 'AMBUJACEM', 'APOLLOHOSP', 'ASIANPAINT', 'ATGL',
       'BAJAJ-AUTO', 'AXISBANK', 'BAJAJFINSV', 'BAJAJHLDNG', 'BAJFINANCE',
       'BANKBARODA', 'BHARTIARTL', 'BEL', 'BHEL', 'BOSCHLTD', 'BPCL',
       'BRITANNIA', 'CANBK', 'CHOLAFIN', 'CIPLA', 'COALINDIA', 'DABUR',
       'DIVISLAB', 'DLF', 'DMART', 'DRREDDY', 'EICHERMOT', 'GAIL',
       'GODREJCP', 'GRASIM', 'HAL', 'HAVELLS', 'HCLTECH', 'HDFCBANK',
       'HDFCLIFE', 'HEROMOTOCO', 'HINDALCO', 'HINDUNILVR', 'ICICIBANK',
       'ICICIGI', 'ICICIPRULI', 'INDUSINDBK', 'INFY', 'IOC', 'IRCTC',
       'IRFC', 'ITC', 'JINDALSTEL', 'JIOFIN', 'JSWENERGY', 'JSWSTEEL',
       'KOTAKBANK', 'LICI', 'LODHA', 'LT', 'LTIM', 'M&M', 'MARUTI',
       'MOTHERSON', 'NAUKRI', 'NESTLEIND', 'NHPC', 'NTPC', 'ONGC', 'PFC',
       'PIDILITIND', 'PNB', 'POWERGRID', 'RECLTD', 'RELIANCE', 'SBILIFE',
       'SBIN', 'SHREECEM', 'SHRIRAMFIN', 'SIEMENS', 'SUNPHARMA',
       'TATACO

In [23]:
# latest_year = 2024

# p_and_l['Year'] = p_and_l['year'].apply(lambda x: clean_period(x))
# def normalize_year(raw: str) -> str:
#     s = str(raw).split(" ")

#     if len(s) > 2:
#         return " ".join(s[:-1])
#     else:
#         s
#     return pd.to_datetime(s,format='%b %Y')
# def 

In [24]:
doc["Year"].unique()

array([2024, 2023, 2022, 2021, 2020, 2019, 2018, 2017, 2016, 2015, 2014,
       2013, 2012, 2011, 2010, 2009, 2008, 2007], dtype=int64)

In [25]:
doc

,id,company_id,Year,Annual_Report
0,1,ABB,2024,https://www.bseindia.com/xml-data/corpfiling/A...
1,2,ABB,2023,https://www.bseindia.com/xml-data/corpfiling/A...
2,3,ABB,2022,https://www.bseindia.com/bseplus/AnnualReport/...
3,4,ABB,2021,https://www.bseindia.com/bseplus/AnnualReport/...
4,5,ABB,2020,https://www.bseindia.com/bseplus/AnnualReport/...
...,...,...,...,...
1580,1581,ZYDUSLIFE,2013,https://www.bseindia.com/bseplus/AnnualReport/...
1581,1582,ZYDUSLIFE,2012,https://www.bseindia.com/bseplus/AnnualReport/...
1582,1583,ZYDUSLIFE,2011,https://www.bseindia.com/bseplus/AnnualReport/...
1583,1584,ZYDUSLIFE,2010,Null


### Excel Loader & Normaliser
loader.py · normalize_year() · normalize_ticker() · 35+ unit tests

In [26]:
p_and_c.columns
doc.columns


Index(['id', 'company_id', 'Year', 'Annual_Report'], dtype='object')

In [27]:
from loader import normalize_year,normalize_ticker

p_and_l['year'] = p_and_l["year"].apply(normalize_year)
p_and_l['company_id'] = p_and_l["company_id"].apply(normalize_ticker)
p_and_c['company_id'] = p_and_c["company_id"].apply(normalize_ticker)
doc['Year'] = doc["Year"].apply(normalize_year)


In [28]:
print(p_and_l["year"].unique())
print("-"*100)
print(doc["Year"].unique())

['Dec 2012' 'Mar 2014' 'Mar 2015' 'Mar 2016' 'Mar 2017' 'Mar 2018'
 'Mar 2019' 'Mar 2020' 'Mar 2021' 'Mar 2022' 'Mar 2023' 'Mar 2024'
 'Dec 2024' 'Mar 2013' 'Dec 2013' 'Dec 2014' 'Dec 2015' 'Dec 2016'
 'Dec 2017' 'Dec 2018' 'Dec 2019' 'Dec 2020' 'Dec 2021' 'Jun 2013'
 'Jun 2014' 'Jun 2015' 'Mar 2011' 'Mar 2012' 'Dec 2022' 'Dec 2023'
 'Sep 2011' 'Sep 2014' 'Sep 2015' 'Sep 2016' 'Sep 2017' 'Sep 2018'
 'Sep 2019' 'Sep 2020' 'Sep 2021' 'Sep 2022' 'Sep 2023' 'Sep 2024']
----------------------------------------------------------------------------------------------------
['2024' '2023' '2022' '2021' '2020' '2019' '2018' '2017' '2016' '2015'
 '2014' '2013' '2012' '2011' '2010' '2009' '2008' '2007']


In [29]:
len(companies_df["id"].unique())


92

In [30]:
companies_df.columns

Index(['id', 'company_logo', 'company_name', 'chart_link', 'about_company',
       'website', 'nse_profile', 'bse_profile', 'face_value', 'book_value',
       'roce_percentage', 'roe_percentage'],
      dtype='object')

Schema Validator — 16 DQ Rules
DQ-01 … DQ-16 · CRITICAL PK/FK · WARNING OPM, balance, sales · generate validation_failures.csv

In [31]:
import os
import sqlite3
BASE_PATH = r"C:\Users\bhrra\Desktop\nifty100"
SCHEMA_PATH = os.path.join(BASE_PATH, "db", "schema.sql")

BASE_PATH = r"C:\Users\bhrra\Desktop\nifty100"

DB_PATH = os.path.join(BASE_PATH, "db", "nifty100.db")
CLEANED_PATH = os.path.join(BASE_PATH, "data", "processed")



In [ ]:




# # Fresh database
# # if os.path.exists(DB_PATH):
# #     os.remove(DB_PATH)

# conn = sqlite3.connect(DB_PATH)

# conn.execute("PRAGMA foreign_keys = ON;")

# with open(SCHEMA_PATH, encoding="utf-8") as f:
#     conn.executescript(f.read())

# conn.commit()

# tables = [
#     "companies",
#     "profitandloss",
#     "balancesheet",
#     "cashflow",
#     "analysis",
#     "documents",
#     "prosandcons",
#     "sectors",
#     "stock_prices",
#     "financial_ratios",
#     "market_cap",
#     "peer_groups"
# ]

# for table in tables:

#     df = pd.read_csv(
#         os.path.join(CLEANED_PATH, f"{table}.csv")
#     )

#     df.to_sql(
#         table,
#         conn,
#         if_exists="append",
#         index=False
#     )

#     print(f"{table}: {len(df)} rows loaded")

# conn.close()
# # ===============================
# schema_sql = """

# CREATE TABLE IF NOT EXISTS companies (
#     id TEXT PRIMARY KEY,
#     company_logo TEXT,
#     company_name TEXT NOT NULL,
#     chart_link TEXT,
#     about_company TEXT,
#     website TEXT,
#     nse_profile TEXT,
#     bse_profile TEXT,
#     face_value REAL NOT NULL,
#     book_value REAL,
#     roce_percentage REAL,
#     roe_percentage REAL
# );

# CREATE TABLE IF NOT EXISTS profitandloss (
#     company_id TEXT NOT NULL,
#     year TEXT NOT NULL,
#     sales REAL NOT NULL,
#     expenses REAL,
#     operating_profit REAL,
#     opm_percentage REAL,
#     other_income REAL,
#     interest REAL,
#     depreciation REAL,
#     profit_before_tax REAL,
#     tax_percentage REAL,
#     net_profit REAL,
#     eps REAL,
#     dividend_payout REAL,
#     PRIMARY KEY (company_id, year),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS balancesheet (
#     company_id TEXT NOT NULL,
#     year TEXT NOT NULL,
#     equity_capital REAL NOT NULL,
#     reserves REAL,
#     borrowings REAL,
#     other_liabilities REAL,
#     total_liabilities REAL NOT NULL,
#     fixed_assets REAL,
#     cwip REAL,
#     investments REAL,
#     other_asset REAL,
#     total_assets REAL NOT NULL,
#     PRIMARY KEY (company_id, year),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS cashflow (
#     company_id TEXT NOT NULL,
#     year TEXT NOT NULL,
#     operating_activity REAL,
#     investing_activity REAL,
#     financing_activity REAL,
#     net_cash_flow REAL,
#     PRIMARY KEY (company_id, year),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS analysis (
#     company_id TEXT PRIMARY KEY,
#     compounded_sales_growth TEXT,
#     compounded_profit_growth TEXT,
#     stock_price_cagr TEXT,
#     roe TEXT,
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS documents (
#     id INTEGER PRIMARY KEY AUTOINCREMENT,
#     company_id TEXT NOT NULL,
#     year INTEGER NOT NULL,
#     annual_report TEXT,
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS prosandcons (
#     id INTEGER PRIMARY KEY AUTOINCREMENT,
#     company_id TEXT NOT NULL,
#     pros TEXT,
#     cons TEXT,
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS sectors (
#     company_id TEXT PRIMARY KEY,
#     broad_sector TEXT NOT NULL,
#     sub_sector TEXT,
#     index_weight_pct REAL,
#     market_cap_category TEXT,
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS stock_prices (
#     company_id TEXT NOT NULL,
#     date TEXT NOT NULL,
#     open_price REAL,
#     high_price REAL,
#     low_price REAL,
#     close_price REAL,
#     volume INTEGER,
#     adjusted_close REAL,
#     PRIMARY KEY (company_id, date),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS financial_ratios (
#     company_id TEXT NOT NULL,
#     year TEXT NOT NULL,
#     net_profit_margin_pct REAL,
#     operating_profit_margin_pct REAL,
#     return_on_equity_pct REAL,
#     debt_to_equity REAL,
#     interest_coverage REAL,
#     asset_turnover REAL,
#     free_cash_flow_cr REAL,
#     capex_cr REAL,
#     earnings_per_share REAL,
#     book_value_per_share REAL,
#     dividend_payout_ratio_pct REAL,
#     total_debt_cr REAL,
#     cash_from_operations_cr REAL,
#     PRIMARY KEY (company_id, year),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS peer_groups (
#     peer_group_name TEXT NOT NULL,
#     company_id TEXT NOT NULL,
#     benchmark TEXT,
#     PRIMARY KEY (peer_group_name, company_id),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS market_cap (
#     company_id TEXT NOT NULL,
#     year INTEGER NOT NULL,
#     market_cap_crore REAL,
#     enterprise_value_crore REAL,
#     pe_ratio REAL,
#     pb_ratio REAL,
#     ev_ebitda REAL,
#     dividend_yield_pct REAL,
#     PRIMARY KEY (company_id, year),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# """

# os.makedirs(r"C:\Users\bhrra\Desktop\nifty100\db", exist_ok=True)

# with open(r"C:\Users\bhrra\Desktop\nifty100\db\schema.sql", "w") as f:
#     f.write(schema_sql)


# DB_PATH = "nifty100.db"

# # Step 1: create/open the db file, enforce FK constraints
# conn = sqlite3.connect(DB_PATH)
# conn.execute("PRAGMA foreign_keys = ON;")

# # Step 2: build all 11 tables from schema.sql
# with open(r"C:\Users\bhrra\Desktop\nifty100\db\schema.sql") as f:
#     conn.executescript(f.read())
# conn.commit()

# # Step 3: insert -- companies FIRST, since every other table's
# # FK depends on it already existing.
# companies_df.to_sql("companies", conn, if_exists="append", index=False)


# analysis_df = pd.read_excel(r"C:\Users\bhrra\Desktop\nifty100\data\raw\analysis.xlsx",skiprows=1)
# cf_df = pd.read_excel(r"C:\Users\bhrra\Desktop\nifty100\data\raw\cashflow.xlsx",skiprows=1)

# sectors_df = pd.read_excel(r"C:\Users\bhrra\Desktop\nifty100\data\raw\supporting datasets\sectors.xlsx")
# stock_prices_df = pd.read_excel(r"C:\Users\bhrra\Desktop\nifty100\data\raw\supporting datasets\stock_prices.xlsx")
# financial_ratios_df = pd.read_excel(r"C:\Users\bhrra\Desktop\nifty100\data\raw\supporting datasets\financial_ratios.xlsx")

# peer_groups_df = pd.read_excel(r"C:\Users\bhrra\Desktop\nifty100\data\raw\supporting datasets\peer_groups.xlsx")
# # Step 4: now the child tables, any order among themselves is fine
# # since they only depend on companies, not on each other.
# p_and_l.to_sql("profitandloss", conn, if_exists="append", index=False)
# bs_df.to_sql("balancesheet", conn, if_exists="append", index=False)
# cf_df.to_sql("cashflow", conn, if_exists="append", index=False)
# analysis_df.to_sql("analysis", conn, if_exists="append", index=False)
# doc.to_sql("documents", conn, if_exists="append", index=False)
# p_and_c.to_sql("prosandcons", conn, if_exists="append", index=False)
# sectors_df.to_sql("sectors", conn, if_exists="append", index=False)
# stock_prices_df.to_sql("stock_prices", conn, if_exists="append", index=False)
# financial_ratios_df.to_sql("financial_ratios", conn, if_exists="append", index=False)
# peer_groups_df.to_sql("peer_groups", conn, if_exists="append", index=False)

# conn.commit()

# # Step 5: verify -- must return zero rows (Day 04/05 exit criteria)
# violations = conn.execute("PRAGMA foreign_key_check;").fetchall()
# print("FK violations:", len(violations))



OperationalError: table companies already exists

In [33]:
comp = pd.read_csv(r"C:\Users\bhrra\Desktop\nifty100\data\processed\companies.csv")

In [34]:
anal = pd.read_csv(r"C:\Users\bhrra\Desktop\nifty100\data\processed\analysis.csv")

In [35]:
from pathlib import Path
import sqlite3
import pandas as pd

# Project folders
BASE_DIR = Path(r"C:\Users\bhrra\Desktop\nifty100")

DATA_DIR = BASE_DIR / "data" / "processed"
DB_PATH = BASE_DIR / "db" / "nifty100.db"

# Check paths
print("Data folder:", DATA_DIR)
print("Data folder exists:", DATA_DIR.exists())

print("\nDatabase:", DB_PATH)
print("Database exists before:", DB_PATH.exists())

# Connect to SQLite
conn = sqlite3.connect(DB_PATH)

# Process every CSV
for file in DATA_DIR.glob("*.csv"):

    print("\n" + "=" * 60)
    print("Processing:", file.name)

    # Table name without .csv
    table_name = file.stem

    print("Table name:", table_name)

    # Read CSV
    df = pd.read_csv(file)

    print("Shape:", df.shape)
    print(df.head(2))

    # Insert into SQLite
    df.to_sql(
        table_name,
        conn,
        if_exists="append",
        index=False
    )

    # Verify immediately
    count = conn.execute(
        f'SELECT COUNT(*) FROM "{table_name}"'
    ).fetchone()[0]

    print(f"Rows in SQLite [{table_name}]: {count:,}")


# Show all tables
print("\n" + "=" * 60)
print("DATABASE SUMMARY")
print("=" * 60)

tables = pd.read_sql_query("""
    SELECT name
    FROM sqlite_master
    WHERE type = 'table'
    ORDER BY name
""", conn)

for table in tables["name"]:

    count = conn.execute(
        f'SELECT COUNT(*) FROM "{table}"'
    ).fetchone()[0]

    print(f"{table:30} → {count:,} rows")

conn.close()

Data folder: C:\Users\bhrra\Desktop\nifty100\data\processed
Data folder exists: True

Database: C:\Users\bhrra\Desktop\nifty100\db\nifty100.db
Database exists before: False

Processing: analysis.csv
Table name: analysis
Shape: (20, 6)
   id company_id compounded_sales_growth compounded_profit_growth  \
0   1   HDFCBANK           10 Years: 21%            10 Years: 22%   
1  10    SBILIFE      5 Years:       24%   5 Years:            6%   

         stock_price_cagr                   roe  
0       10 Years:     15%     10 Years:     17%  
1       5 Years:       8%  5 Years          14%  
Rows in SQLite [analysis]: 20

Processing: balancesheet.csv
Table name: balancesheet
Shape: (1312, 13)
    id company_id      year  equity_capital  reserves  borrowings  \
0  136        ABB  Dec 2012            21.0       626           0   
1  137        ABB  Mar 2014            21.0       767           0   

   other_liabilities  total_liabilities  fixed_assets  cwip  investments  \
0                260

Rows in SQLite [companies]: 89

Processing: documents.csv
Table name: documents
Shape: (1533, 4)
   id company_id  Year                                      Annual_Report
0   1        ABB  2024  https://www.bseindia.com/xml-data/corpfiling/A...
1   2        ABB  2023  https://www.bseindia.com/xml-data/corpfiling/A...
Rows in SQLite [documents]: 1,533

Processing: financial_ratios.csv
Table name: financial_ratios
Shape: (1085, 16)
   id company_id      year  net_profit_margin_pct  \
0   6        ABB  Mar 2016                   9.76   
1   7        ABB  Mar 2016                   9.76   

   operating_profit_margin_pct  return_on_equity_pct  debt_to_equity  \
0                         14.0                 21.34             0.0   
1                         14.0                 21.34             0.0   

   interest_coverage  asset_turnover  free_cash_flow_cr  capex_cr  \
0           138.3333          1.6176              172.0      77.0   
1           138.3333          1.6176              7

In [ ]:
# from pathlib import Path
# import sqlite3
# import pandas as pd


# # ============================================================
# # 1. PATHS
# # ============================================================

# BASE_DIR = Path(r"C:\Users\bhrra\Desktop\nifty100")

# DATA_DIR = BASE_DIR / "data" / "processed"
# DB_PATH = BASE_DIR / "db" / "nifty100.db"

# print("Data folder:", DATA_DIR)
# print("Database:", DB_PATH)
# print("Data folder exists:", DATA_DIR.exists())


# # ============================================================
# # 2. CONNECT TO DATABASE
# # ============================================================

# DB_PATH.parent.mkdir(parents=True, exist_ok=True)

# conn = sqlite3.connect(DB_PATH)

# # Enable foreign keys
# conn.execute("PRAGMA foreign_keys = ON;")


# # ============================================================
# # 3. CREATE TABLES
# # ============================================================

# schema = """

# CREATE TABLE IF NOT EXISTS companies (
#     id TEXT PRIMARY KEY,
#     company_logo TEXT,
#     company_name TEXT NOT NULL,
#     chart_link TEXT,
#     about_company TEXT,
#     website TEXT,
#     nse_profile TEXT,
#     bse_profile TEXT,
#     face_value REAL NOT NULL,
#     book_value REAL,
#     roce_percentage REAL,
#     roe_percentage REAL
# );

# CREATE TABLE IF NOT EXISTS profitandloss (
#     company_id TEXT NOT NULL,
#     year TEXT NOT NULL,
#     sales REAL NOT NULL,
#     expenses REAL,
#     operating_profit REAL,
#     opm_percentage REAL,
#     other_income REAL,
#     interest REAL,
#     depreciation REAL,
#     profit_before_tax REAL,
#     tax_percentage REAL,
#     net_profit REAL,
#     eps REAL,
#     dividend_payout REAL,
#     PRIMARY KEY (company_id, year),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS balancesheet (
#     company_id TEXT NOT NULL,
#     year TEXT NOT NULL,
#     equity_capital REAL NOT NULL,
#     reserves REAL,
#     borrowings REAL,
#     other_liabilities REAL,
#     total_liabilities REAL NOT NULL,
#     fixed_assets REAL,
#     cwip REAL,
#     investments REAL,
#     other_asset REAL,
#     total_assets REAL NOT NULL,
#     PRIMARY KEY (company_id, year),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS cashflow (
#     company_id TEXT NOT NULL,
#     year TEXT NOT NULL,
#     operating_activity REAL,
#     investing_activity REAL,
#     financing_activity REAL,
#     net_cash_flow REAL,
#     PRIMARY KEY (company_id, year),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS analysis (
#     company_id TEXT PRIMARY KEY,
#     compounded_sales_growth TEXT,
#     compounded_profit_growth TEXT,
#     stock_price_cagr TEXT,
#     roe TEXT,
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS documents (
#     id INTEGER PRIMARY KEY AUTOINCREMENT,
#     company_id TEXT NOT NULL,
#     year INTEGER NOT NULL,
#     annual_report TEXT,
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS prosandcons (
#     id INTEGER PRIMARY KEY AUTOINCREMENT,
#     company_id TEXT NOT NULL,
#     pros TEXT,
#     cons TEXT,
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS sectors (
#     company_id TEXT PRIMARY KEY,
#     broad_sector TEXT NOT NULL,
#     sub_sector TEXT,
#     index_weight_pct REAL,
#     market_cap_category TEXT,
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS stock_prices (
#     company_id TEXT NOT NULL,
#     date TEXT NOT NULL,
#     open_price REAL,
#     high_price REAL,
#     low_price REAL,
#     close_price REAL,
#     volume INTEGER,
#     adjusted_close REAL,
#     PRIMARY KEY (company_id, date),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS financial_ratios (
#     company_id TEXT NOT NULL,
#     year TEXT NOT NULL,
#     net_profit_margin_pct REAL,
#     operating_profit_margin_pct REAL,
#     return_on_equity_pct REAL,
#     debt_to_equity REAL,
#     interest_coverage REAL,
#     asset_turnover REAL,
#     free_cash_flow_cr REAL,
#     capex_cr REAL,
#     earnings_per_share REAL,
#     book_value_per_share REAL,
#     dividend_payout_ratio_pct REAL,
#     total_debt_cr REAL,
#     cash_from_operations_cr REAL,
#     PRIMARY KEY (company_id, year),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS peer_groups (
#     peer_group_name TEXT NOT NULL,
#     company_id TEXT NOT NULL,
#     benchmark TEXT,
#     PRIMARY KEY (peer_group_name, company_id),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# CREATE TABLE IF NOT EXISTS market_cap (
#     company_id TEXT NOT NULL,
#     year INTEGER NOT NULL,
#     market_cap_crore REAL,
#     enterprise_value_crore REAL,
#     pe_ratio REAL,
#     pb_ratio REAL,
#     ev_ebitda REAL,
#     dividend_yield_pct REAL,
#     PRIMARY KEY (company_id, year),
#     FOREIGN KEY (company_id) REFERENCES companies(id)
# );

# """

# conn.executescript(schema)
# conn.commit()

# print("\nAll tables created successfully.")


# # ============================================================
# # 4. CSV → TABLE MAPPING
# # ============================================================

# tables = [
#     "companies",
#     "profitandloss",
#     "balancesheet",
#     "cashflow",
#     "analysis",
#     "documents",
#     "prosandcons",
#     "sectors",
#     "stock_prices",
#     "financial_ratios",
#     "peer_groups",
#     "market_cap"
# ]


# # ============================================================
# # 5. INSERT DATA
# # ============================================================

# for table_name in tables:

#     csv_path = DATA_DIR / f"{table_name}.csv"

#     if not csv_path.exists():
#         print(f"\n⚠️ {table_name}.csv not found")
#         continue

#     print("\n" + "=" * 60)
#     print(f"Processing: {table_name}.csv")

#     df = pd.read_csv(csv_path)

#     print("CSV rows:", len(df))
#     print("CSV columns:", list(df.columns))

#     # --------------------------------------------------------
#     # Remove duplicate primary-key rows where necessary
#     # --------------------------------------------------------

#     if table_name in [
#         "profitandloss",
#         "balancesheet",
#         "cashflow",
#         "financial_ratios",
#         "stock_prices",
#         "market_cap",
#         "peer_groups"
#     ]:

#         if table_name == "stock_prices":
#             subset = ["company_id", "date"]

#         elif table_name == "peer_groups":
#             subset = ["peer_group_name", "company_id"]

#         else:
#             subset = ["company_id", "year"]

#         before = len(df)

#         df = df.drop_duplicates(
#             subset=subset,
#             keep="last"
#         )

#         after = len(df)

#         print(
#             f"Duplicates removed: {before - after}"
#         )

#     # --------------------------------------------------------
#     # Get actual database columns
#     # --------------------------------------------------------

#     db_columns = pd.read_sql_query(
#         f'PRAGMA table_info("{table_name}")',
#         conn
#     )["name"].tolist()

#     # Only keep columns that exist in DB
#     common_columns = [
#         col for col in df.columns
#         if col in db_columns
#     ]

#     df = df[common_columns]

#     print("Columns inserted:", common_columns)

#     # --------------------------------------------------------
#     # Insert
#     # --------------------------------------------------------

#     try:

#         df.to_sql(
#             table_name,
#             conn,
#             if_exists="append",
#             index=False
#         )

#         print(
#             f"✅ Inserted {len(df):,} rows into {table_name}"
#         )

#     except Exception as e:

#         print(
#             f"❌ Error inserting {table_name}:"
#         )

#         print(e)


# # ============================================================
# # 6. VERIFY DATABASE
# # ============================================================

# print("\n\n")
# print("=" * 70)
# print("DATABASE SUMMARY")
# print("=" * 70)

# db_tables = pd.read_sql_query("""
#     SELECT name
#     FROM sqlite_master
#     WHERE type = 'table'
#     ORDER BY name
# """, conn)

# for table in db_tables["name"]:

#     count = conn.execute(
#         f'SELECT COUNT(*) FROM "{table}"'
#     ).fetchone()[0]

#     print(
#         f"{table:25} → {count:,} rows"
#     )


# # ============================================================
# # 7. FOREIGN KEY CHECK
# # ============================================================

# print("\nForeign key check:")

# fk_errors = conn.execute(
#     "PRAGMA foreign_key_check;"
# ).fetchall()

# if fk_errors:
#     print("❌ Foreign key errors:")
#     for error in fk_errors:
#         print(error)
# else:
#     print("✅ No foreign key errors.")


# # ============================================================
# # 8. CLOSE
# # ============================================================

# conn.close()

# print("\nDatabase loading completed.")

Data folder: C:\Users\bhrra\Desktop\nifty100\data\processed
Database: C:\Users\bhrra\Desktop\nifty100\db\nifty100.db
Data folder exists: True

All tables created successfully.

Processing: companies.csv
CSV rows: 89
CSV columns: ['id', 'company_logo', 'company_name', 'chart_link', 'about_company', 'website', 'nse_profile', 'bse_profile', 'face_value', 'book_value', 'roce_percentage', 'roe_percentage']
Columns inserted: ['id', 'company_logo', 'company_name', 'chart_link', 'about_company', 'website', 'nse_profile', 'bse_profile', 'face_value', 'book_value', 'roce_percentage', 'roe_percentage']
✅ Inserted 89 rows into companies

Processing: profitandloss.csv
CSV rows: 1156
CSV columns: ['id', 'company_id', 'year', 'sales', 'expenses', 'operating_profit', 'opm_percentage', 'other_income', 'interest', 'depreciation', 'profit_before_tax', 'tax_percentage', 'net_profit', 'eps', 'dividend_payout']
Duplicates removed: 12
Columns inserted: ['id', 'company_id', 'year', 'sales', 'expenses', 'opera

In [37]:
from pathlib import Path
import sqlite3
import pandas as pd


# ============================================================
# 1. PATHS
# ============================================================

BASE_DIR = Path(r"C:\Users\bhrra\Desktop\nifty100")

DATA_DIR = BASE_DIR / "data" / "processed"
DB_PATH = BASE_DIR / "db" / "nifty100.db"

print("Data folder:", DATA_DIR)
print("Database:", DB_PATH)
print("Data folder exists:", DATA_DIR.exists())


# ============================================================
# 2. CONNECT TO DATABASE
# ============================================================

DB_PATH.parent.mkdir(parents=True, exist_ok=True)

conn = sqlite3.connect(DB_PATH)

# Enable foreign keys
conn.execute("PRAGMA foreign_keys = ON;")


# ============================================================
# 3. CREATE TABLES
# ============================================================

schema = """

CREATE TABLE IF NOT EXISTS companies (
    id TEXT PRIMARY KEY,
    company_logo TEXT,
    company_name TEXT NOT NULL,
    chart_link TEXT,
    about_company TEXT,
    website TEXT,
    nse_profile TEXT,
    bse_profile TEXT,
    face_value REAL NOT NULL,
    book_value REAL,
    roce_percentage REAL,
    roe_percentage REAL
);

CREATE TABLE IF NOT EXISTS profitandloss (
    company_id TEXT NOT NULL,
    year TEXT NOT NULL,
    sales REAL NOT NULL,
    expenses REAL,
    operating_profit REAL,
    opm_percentage REAL,
    other_income REAL,
    interest REAL,
    depreciation REAL,
    profit_before_tax REAL,
    tax_percentage REAL,
    net_profit REAL,
    eps REAL,
    dividend_payout REAL,
    PRIMARY KEY (company_id, year),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

CREATE TABLE IF NOT EXISTS balancesheet (
    company_id TEXT NOT NULL,
    year TEXT NOT NULL,
    equity_capital REAL NOT NULL,
    reserves REAL,
    borrowings REAL,
    other_liabilities REAL,
    total_liabilities REAL NOT NULL,
    fixed_assets REAL,
    cwip REAL,
    investments REAL,
    other_asset REAL,
    total_assets REAL NOT NULL,
    PRIMARY KEY (company_id, year),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

CREATE TABLE IF NOT EXISTS cashflow (
    company_id TEXT NOT NULL,
    year TEXT NOT NULL,
    operating_activity REAL,
    investing_activity REAL,
    financing_activity REAL,
    net_cash_flow REAL,
    PRIMARY KEY (company_id, year),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

CREATE TABLE IF NOT EXISTS analysis (
    company_id TEXT PRIMARY KEY,
    compounded_sales_growth TEXT,
    compounded_profit_growth TEXT,
    stock_price_cagr TEXT,
    roe TEXT,
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

CREATE TABLE IF NOT EXISTS documents (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    company_id TEXT NOT NULL,
    year INTEGER NOT NULL,
    annual_report TEXT,
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

CREATE TABLE IF NOT EXISTS prosandcons (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    company_id TEXT NOT NULL,
    pros TEXT,
    cons TEXT,
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

CREATE TABLE IF NOT EXISTS sectors (
    company_id TEXT PRIMARY KEY,
    broad_sector TEXT NOT NULL,
    sub_sector TEXT,
    index_weight_pct REAL,
    market_cap_category TEXT,
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

CREATE TABLE IF NOT EXISTS stock_prices (
    company_id TEXT NOT NULL,
    date TEXT NOT NULL,
    open_price REAL,
    high_price REAL,
    low_price REAL,
    close_price REAL,
    volume INTEGER,
    adjusted_close REAL,
    PRIMARY KEY (company_id, date),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

CREATE TABLE IF NOT EXISTS financial_ratios (
    company_id TEXT NOT NULL,
    year TEXT NOT NULL,
    net_profit_margin_pct REAL,
    operating_profit_margin_pct REAL,
    return_on_equity_pct REAL,
    debt_to_equity REAL,
    interest_coverage REAL,
    asset_turnover REAL,
    free_cash_flow_cr REAL,
    capex_cr REAL,
    earnings_per_share REAL,
    book_value_per_share REAL,
    dividend_payout_ratio_pct REAL,
    total_debt_cr REAL,
    cash_from_operations_cr REAL,
    PRIMARY KEY (company_id, year),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

CREATE TABLE IF NOT EXISTS peer_groups (
    peer_group_name TEXT NOT NULL,
    company_id TEXT NOT NULL,
    benchmark TEXT,
    PRIMARY KEY (peer_group_name, company_id),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

CREATE TABLE IF NOT EXISTS market_cap (
    company_id TEXT NOT NULL,
    year INTEGER NOT NULL,
    market_cap_crore REAL,
    enterprise_value_crore REAL,
    pe_ratio REAL,
    pb_ratio REAL,
    ev_ebitda REAL,
    dividend_yield_pct REAL,
    PRIMARY KEY (company_id, year),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

"""

conn.executescript(schema)
conn.commit()

print("\nAll tables created successfully.")


# ============================================================
# 4. CSV → TABLE MAPPING
# ============================================================

tables = [
    "companies",
    "profitandloss",
    "balancesheet",
    "cashflow",
    "analysis",
    "documents",
    "prosandcons",
    "sectors",
    "stock_prices",
    "financial_ratios",
    "peer_groups",
    "market_cap"
]


# ============================================================
# 5. INSERT DATA
# ============================================================

for table_name in tables:

    csv_path = DATA_DIR / f"{table_name}.csv"

    if not csv_path.exists():
        print(f"\n⚠️ {table_name}.csv not found")
        continue

    print("\n" + "=" * 60)
    print(f"Processing: {table_name}.csv")

    df = pd.read_csv(csv_path)

    print("CSV rows:", len(df))
    print("CSV columns:", list(df.columns))

    # --------------------------------------------------------
    # Remove duplicate primary-key rows where necessary
    # --------------------------------------------------------

    if table_name in [
        "profitandloss",
        "balancesheet",
        "cashflow",
        "financial_ratios",
        "stock_prices",
        "market_cap",
        "peer_groups"
    ]:

        if table_name == "stock_prices":
            subset = ["company_id", "date"]

        elif table_name == "peer_groups":
            subset = ["peer_group_name", "company_id"]

        else:
            subset = ["company_id", "year"]

        before = len(df)

        df = df.drop_duplicates(
            subset=subset,
            keep="last"
        )

        after = len(df)

        print(
            f"Duplicates removed: {before - after}"
        )

    # --------------------------------------------------------
    # Get actual database columns
    # --------------------------------------------------------

    db_columns = pd.read_sql_query(
        f'PRAGMA table_info("{table_name}")',
        conn
    )["name"].tolist()

    # Only keep columns that exist in DB
    common_columns = [
        col for col in df.columns
        if col in db_columns
    ]

    df = df[common_columns]

    print("Columns inserted:", common_columns)

    # --------------------------------------------------------
    # Insert
    # --------------------------------------------------------

    try:

        df.to_sql(
            table_name,
            conn,
            if_exists="append",
            index=False
        )

        print(
            f"✅ Inserted {len(df):,} rows into {table_name}"
        )

    except Exception as e:

        print(
            f"❌ Error inserting {table_name}:"
        )

        print(e)


# ============================================================
# 6. VERIFY DATABASE
# ============================================================

print("\n\n")
print("=" * 70)
print("DATABASE SUMMARY")
print("=" * 70)

db_tables = pd.read_sql_query("""
    SELECT name
    FROM sqlite_master
    WHERE type = 'table'
    ORDER BY name
""", conn)

for table in db_tables["name"]:

    count = conn.execute(
        f'SELECT COUNT(*) FROM "{table}"'
    ).fetchone()[0]

    print(
        f"{table:25} → {count:,} rows"
    )


# ============================================================
# 7. FOREIGN KEY CHECK
# ============================================================

print("\nForeign key check:")

fk_errors = conn.execute(
    "PRAGMA foreign_key_check;"
).fetchall()

if fk_errors:
    print("❌ Foreign key errors:")
    for error in fk_errors:
        print(error)
else:
    print("✅ No foreign key errors.")


# ============================================================
# 8. CLOSE
# ============================================================

conn.close()

print("\nDatabase loading completed.")

Data folder: C:\Users\bhrra\Desktop\nifty100\data\processed
Database: C:\Users\bhrra\Desktop\nifty100\db\nifty100.db
Data folder exists: True

All tables created successfully.

Processing: companies.csv
CSV rows: 89
CSV columns: ['id', 'company_logo', 'company_name', 'chart_link', 'about_company', 'website', 'nse_profile', 'bse_profile', 'face_value', 'book_value', 'roce_percentage', 'roe_percentage']
Columns inserted: ['id', 'company_logo', 'company_name', 'chart_link', 'about_company', 'website', 'nse_profile', 'bse_profile', 'face_value', 'book_value', 'roce_percentage', 'roe_percentage']
✅ Inserted 89 rows into companies

Processing: profitandloss.csv
CSV rows: 1156
CSV columns: ['id', 'company_id', 'year', 'sales', 'expenses', 'operating_profit', 'opm_percentage', 'other_income', 'interest', 'depreciation', 'profit_before_tax', 'tax_percentage', 'net_profit', 'eps', 'dividend_payout']
Duplicates removed: 12
Columns inserted: ['id', 'company_id', 'year', 'sales', 'expenses', 'opera

In [38]:
conn = sqlite3.connect(DB_PATH)

In [39]:
conn.close()

In [40]:
# from pathlib import Path

# DB_PATH = Path(r"C:\Users\bhrra\Desktop\nifty100\db\nifty100.db")

# if DB_PATH.exists():
#     DB_PATH.unlink()
#     print("Database deleted successfully.")
# else:
#     print("Database does not exist.")

In [48]:
from pathlib import Path
import sqlite3
import pandas as pd

DB_PATH = Path(r"C:\Users\bhrra\Desktop\nifty100\db\nifty100.db")

print("Using database:")
print(DB_PATH)

print("Exists:", DB_PATH.exists())

conn = sqlite3.connect(DB_PATH)

tables = pd.read_sql_query("""
    SELECT name
    FROM sqlite_master
    WHERE type = 'table'
    ORDER BY name
""", conn)

print("\nTables:")
print(tables)

for table in tables["name"]:
    count = conn.execute(
        f'SELECT COUNT(*) FROM "{table}"'
    ).fetchone()[0]

    print(f"{table:30} → {count:,} rows")

conn.close()

Using database:
C:\Users\bhrra\Desktop\nifty100\db\nifty100.db
Exists: True

Tables:
                name
0           analysis
1       balancesheet
2           cashflow
3          companies
4          documents
5   financial_ratios
6         market_cap
7        peer_groups
8      profitandloss
9        prosandcons
10           sectors
11      stock_prices
analysis                       → 60 rows
balancesheet                   → 3,762 rows
cashflow                       → 3,509 rows
companies                      → 267 rows
documents                      → 4,599 rows
financial_ratios               → 3,117 rows
market_cap                     → 1,656 rows
peer_groups                    → 168 rows
profitandloss                  → 3,444 rows
prosandcons                    → 30 rows
sectors                        → 276 rows
stock_prices                   → 16,560 rows


In [42]:
conn = sqlite3.connect("nifty100.db")

anal.to_sql(
    "analysis",
    conn,
    if_exists="append",
    index=False
)

20

In [43]:
conn = sqlite3.connect("nifty100.db")

comp.to_sql(
    "companies",
    conn,
    if_exists="append",
    index=False
)

89

In [44]:
# conn.execute("DELETE FROM companies")
# conn.commit()

In [ ]:
print(len(tables))


1156


In [50]:
# conn = sqlite3.connect("nifty100.db")


In [55]:
import sqlite3
import pandas as pd

# Connect
conn = sqlite3.connect(DB_PATH)

queries = {
    "row_counts": """
        SELECT 'companies' AS table_name, COUNT(*) AS rows FROM companies
        UNION ALL SELECT 'profitandloss', COUNT(*) FROM profitandloss
        UNION ALL SELECT 'balancesheet', COUNT(*) FROM balancesheet
        UNION ALL SELECT 'cashflow', COUNT(*) FROM cashflow
        UNION ALL SELECT 'stock_prices', COUNT(*) FROM stock_prices;
    """,

    "company_count": """
        SELECT COUNT(*) AS company_count
        FROM companies;
    """,

    "year_coverage": """
        SELECT
            company_id,
            MIN(year) AS first_year,
            MAX(year) AS last_year,
            COUNT(DISTINCT year) AS years_available
        FROM profitandloss
        GROUP BY company_id
        ORDER BY years_available;
    """,

    "companies_fewer_than_5_years": """
        SELECT
            company_id,
            COUNT(DISTINCT year) AS years_available
        FROM profitandloss
        GROUP BY company_id
        HAVING years_available < 5;
    """,

    "duplicate_company_year": """
        SELECT
            company_id,
            year,
            COUNT(*) AS occurrences
        FROM profitandloss
        GROUP BY company_id, year
        HAVING occurrences > 1;
    """,

    "orphan_profitandloss_company_ids": """
        SELECT p.company_id
        FROM profitandloss p
        LEFT JOIN companies c ON p.company_id = c.id
        WHERE c.id IS NULL;
    """,

    "profitandloss_null_counts": """
        SELECT
            SUM(CASE WHEN sales IS NULL THEN 1 ELSE 0 END) AS null_sales,
            SUM(CASE WHEN net_profit IS NULL THEN 1 ELSE 0 END) AS null_net_profit,
            SUM(CASE WHEN eps IS NULL THEN 1 ELSE 0 END) AS null_eps
        FROM profitandloss;
    """,

    "sector_distribution": """
        SELECT
            broad_sector,
            COUNT(*) AS company_count
        FROM sectors
        GROUP BY broad_sector
        ORDER BY company_count DESC;
    """,

    "balance_sheet_imbalance": """
        SELECT
            company_id,
            year,
            total_assets,
            total_liabilities,
            ROUND(ABS(total_assets - total_liabilities) * 100.0 / total_assets, 2) AS diff_pct
        FROM balancesheet
        WHERE ABS(total_assets - total_liabilities) * 1.0 / total_assets >= 0.01
        ORDER BY diff_pct DESC;
    """,

    "negative_or_zero_sales": """
        SELECT
            company_id,
            year,
            sales
        FROM profitandloss
        WHERE sales <= 0;
    """
}
for name, query in queries.items():
    print(f"\n{name}")
    df = pd.read_sql_query(query, conn)
    display(df)
conn.close()


row_counts


,table_name,rows
0,companies,267
1,profitandloss,3444
2,balancesheet,3762
3,cashflow,3509
4,stock_prices,16560



company_count


,company_count
0,267



year_coverage


,company_id,first_year,last_year,years_available
0,JIOFIN,Mar 2024,Mar 2024,1
1,LICI,Mar 2021,Mar 2024,4
2,ATGL,Mar 2018,Mar 2024,7
3,ZOMATO,Mar 2018,Mar 2024,7
4,ADANIGREEN,Mar 2017,Mar 2024,8
...,...,...,...,...
94,UNITDSPR,Mar 2013,Mar 2024,12
95,VBL,Dec 2012,Dec 2023,12
96,VEDL,Mar 2013,Mar 2024,12
97,WIPRO,Mar 2013,Mar 2024,12



companies_fewer_than_5_years


,company_id,years_available
0,JIOFIN,1
1,LICI,4



duplicate_company_year


,company_id,year,occurrences
0,ABB,Dec 2012,3
1,ABB,Mar 2014,3
2,ABB,Mar 2015,3
3,ABB,Mar 2016,3
4,ABB,Mar 2017,3
...,...,...,...
1139,ZYDUSLIFE,Mar 2020,3
1140,ZYDUSLIFE,Mar 2021,3
1141,ZYDUSLIFE,Mar 2022,3
1142,ZYDUSLIFE,Mar 2023,3



orphan_profitandloss_company_ids


,company_id
0,LTIM
1,LTIM
2,LTIM
3,LTIM
4,LTIM
...,...
376,INDIGO
377,INDIGO
378,INDIGO
379,INDIGO



profitandloss_null_counts


,null_sales,null_net_profit,null_eps
0,0,0,0



sector_distribution


,broad_sector,company_count
0,Financials,69
1,Energy,42
2,Consumer Discretionary,42
3,Industrials,30
4,Materials,27
5,Consumer Staples,21
6,Healthcare,18
7,Information Technology,15
8,Real Estate,6
9,Communication Services,6



balance_sheet_imbalance


,company_id,year,total_assets,total_liabilities,diff_pct



negative_or_zero_sales


,company_id,year,sales


In [ ]:
p_and_l

,id,company_id,year,sales,expenses,operating_profit,opm_percentage,other_income,interest,depreciation,profit_before_tax,tax_percentage,net_profit,eps,dividend_payout
0,61,ABB,Dec 2012,1653,1451,202.0,12.0,33,0,19,215,33.0,145,68.0,25.0
1,62,ABB,Mar 2014,2276,2009,267.0,12.0,49,0,22,295,33.0,198,93.0,25.0
2,63,ABB,Mar 2015,2289,1977,312.0,14.0,48,0,15,344,34.0,229,108.0,29.0
3,64,ABB,Mar 2016,2614,2250,365.0,14.0,50,3,14,398,36.0,255,120.0,29.0
4,65,ABB,Mar 2017,2903,2505,398.0,14.0,57,2,16,436,37.0,277,130.0,31.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1271,1370,INDIGO,Mar 2021,14641,2170,14638.0,2.0,0,1036,4699,-5830,0.0,-6171,-151.0,0.0
1272,1371,INDIGO,Mar 2022,25931,2386,25373.0,558.0,2,725,5068,-6171,0.0,-317,-160.0,0.0
1273,1372,INDIGO,Mar 2023,54446,3168,47926.0,6521.0,12,1431,5101,-317,0.0,-317,-8.0,0.0
1274,1373,INDIGO,Mar 2024,68904,4805,52573.0,16331.0,24,2326,6406,8043,-2.0,8167,212.0,0.0


In [ ]:
path = Path(r"C:\Users\bhrra\Desktop\nifty100\notebooks\exploratory_queries.sql")

with open(path, "w", encoding="utf-8") as file:
    for name, query in queries.items():

        file.write(f"-- {name}\n")
        file.write(query.strip())
        file.write("\n\n")

In [ ]:
companies_df[companies_df["face_value"].isna()]

,id,company_logo,company_name,chart_link,about_company,website,nse_profile,bse_profile,face_value,book_value,roce_percentage,roe_percentage
91,TVSMOTOR,https://mkt.in/static/mkt-icons/nifty100/TVSMO...,TVS Motor Company Ltd,https://in.tradingview.com/chart/qGsydD2w/?sym...,'TVS Motor Company Ltd (TVSM) is engaged in ma...,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
import sqlite3

conn = sqlite3.connect("nifty100.db")
conn.execute("PRAGMA foreign_keys = OFF")

tables = conn.execute("""
    SELECT name
    FROM sqlite_master
    WHERE type = 'table'
      AND name NOT LIKE 'sqlite_%'
""").fetchall()

for (table,) in tables:
    conn.execute(f'DROP TABLE "{table}"')

conn.commit()

conn.execute("PRAGMA foreign_keys = ON")

In [ ]:
import pandas as pd
from pathlib import Path

RAW_DIR = Path(r"C:\Users\bhrra\Desktop\nifty100\data\raw")
SUPP_DIR = RAW_DIR / "supporting datasets"
OUT_DIR = Path(r"C:\Users\bhrra\Desktop\nifty100\output")
OUT_DIR.mkdir(exist_ok=True)

CORE_FILES = ["companies", "profitandloss", "balancesheet", "cashflow",
              "analysis", "documents", "prosandcons"]
SUPP_FILES = ["sectors", "stock_prices", "market_cap",
              "financial_ratios", "peer_groups"]

def split_nulls(name: str, path: Path, header_row: int):
    df = pd.read_excel(path, header=header_row)

    null_mask = df.isnull().any(axis=1)     # True if ANY column is null in that row
    clean_df = df[~null_mask]
    null_df = df[null_mask]

    clean_df.to_csv(OUT_DIR / f"{name}.csv", index=False)
    null_df.to_csv(OUT_DIR / f"{name}_null_values.csv", index=False)

    print(f"{name}: {len(df)} rows -> {len(clean_df)} clean, {len(null_df)} null")

for name in CORE_FILES:
    split_nulls(name, RAW_DIR / f"{name}.xlsx", header_row=1)

for name in SUPP_FILES:
    split_nulls(name, SUPP_DIR / f"{name}.xlsx", header_row=0)

companies: 92 rows -> 89 clean, 3 null


profitandloss: 1276 rows -> 1156 clean, 120 null
balancesheet: 1312 rows -> 1312 clean, 0 null
cashflow: 1187 rows -> 1185 clean, 2 null
analysis: 20 rows -> 20 clean, 0 null
documents: 1585 rows -> 1533 clean, 52 null
prosandcons: 16 rows -> 10 clean, 6 null
sectors: 92 rows -> 92 clean, 0 null
stock_prices: 5520 rows -> 5520 clean, 0 null
market_cap: 552 rows -> 552 clean, 0 null
financial_ratios: 1184 rows -> 1085 clean, 99 null
peer_groups: 56 rows -> 56 clean, 0 null


In [57]:
import sqlite3

conn = sqlite3.connect(DB_PATH)

tables = conn.execute("""
    SELECT name
    FROM sqlite_master
    WHERE type = 'table'
    AND name NOT LIKE 'sqlite_%';
""").fetchall()

print([table[0] for table in tables])
conn.close()

['analysis', 'balancesheet', 'cashflow', 'companies', 'documents', 'financial_ratios', 'market_cap', 'peer_groups', 'profitandloss', 'prosandcons', 'sectors', 'stock_prices']


In [ ]:
import os
import shutil

source = r"C:\Users\bhrra\Desktop\nifty100\output"

cleaned = os.path.join(source, "cleaned")
null_values = os.path.join(source, "null_values")

os.makedirs(cleaned, exist_ok=True)
os.makedirs(null_values, exist_ok=True)

for file in os.listdir(source):

    file_path = os.path.join(source, file)

    if not os.path.isfile(file_path):
        continue

    if "_null_values" in file:
        shutil.move(file_path, os.path.join(null_values, file))
    elif file.endswith(".xlsx") or file.endswith(".csv"):
        shutil.move(file_path, os.path.join(cleaned, file))

SQLite Database Schema
schema.sql · 10 tables · PK/FK · loader.py · PRAGMA foreign_keys = ON

In [ ]:
sql_schema = {
	"""CREATE TABLE companies (
    id                TEXT PRIMARY KEY,
    company_logo      TEXT,
    company_name      TEXT NOT NULL,
    chart_link        TEXT,
    about_company     TEXT,
    website           TEXT,
    nse_profile       TEXT,
    bse_profile       TEXT,
    face_value        REAL NOT NULL,
    book_value        REAL,
    roce_percentage   REAL,
    roe_percentage    REAL
);""",
"""
CREATE TABLE profitandloss (
    company_id         TEXT NOT NULL,
    year                TEXT NOT NULL,
    sales                REAL NOT NULL,
    expenses             REAL,
    operating_profit     REAL,
    opm_percentage       REAL,
    other_income         REAL,
    interest             REAL,
    depreciation         REAL,
    profit_before_tax    REAL,
    tax_percentage       REAL,
    net_profit           REAL,
    eps                  REAL,
    dividend_payout      REAL,
    PRIMARY KEY (company_id, year),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);
""",
"""
CREATE TABLE balancesheet (
    company_id         TEXT NOT NULL,
    year                TEXT NOT NULL,
    equity_capital       REAL NOT NULL,
    reserves              REAL,
    borrowings            REAL,
    other_liabilities     REAL,
    total_liabilities     REAL NOT NULL,
    fixed_assets          REAL,
    cwip                  REAL,
    investments           REAL,
    other_asset           REAL,
    total_assets          REAL NOT NULL,
    PRIMARY KEY (company_id, year),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);
"""
,
"""
CREATE TABLE cashflow (
    company_id          TEXT NOT NULL,
    year                 TEXT NOT NULL,
    operating_activity    REAL,
    investing_activity    REAL,
    financing_activity    REAL,
    net_cash_flow         REAL,
    PRIMARY KEY (company_id, year),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);
"""
,
"""
CREATE TABLE analysis (
    company_id                 TEXT PRIMARY KEY,
    compounded_sales_growth    TEXT,
    compounded_profit_growth   TEXT,
    stock_price_cagr           TEXT,
    roe                        TEXT,
    FOREIGN KEY (company_id) REFERENCES companies(id)
);
"""
,
"""
CREATE TABLE documents (
    id             INTEGER PRIMARY KEY AUTOINCREMENT,
    company_id     TEXT NOT NULL,
    year           INTEGER NOT NULL,
    annual_report  TEXT,
    FOREIGN KEY (company_id) REFERENCES companies(id)
);
"""
,
"""
CREATE TABLE prosandcons (
    id          INTEGER PRIMARY KEY AUTOINCREMENT,
    company_id  TEXT NOT NULL,
    pros        TEXT,
    cons        TEXT,
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

"""
,
"""
CREATE TABLE sectors (
    company_id           TEXT PRIMARY KEY,
    broad_sector          TEXT NOT NULL,
    sub_sector            TEXT,
    index_weight_pct      REAL,
    market_cap_category   TEXT,
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

"""
,
"""CREATE TABLE stock_prices (
    company_id       TEXT NOT NULL,
    date              TEXT NOT NULL,
    open_price         REAL,
    high_price         REAL,
    low_price          REAL,
    close_price        REAL,
    volume             INTEGER,
    adjusted_close     REAL,
    PRIMARY KEY (company_id, date),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);

"""
,
"""CREATE TABLE financial_ratios (
    company_id                     TEXT NOT NULL,
    year                            TEXT NOT NULL,
    net_profit_margin_pct            REAL,
    operating_profit_margin_pct      REAL,
    return_on_equity_pct             REAL,
    debt_to_equity                   REAL,
    interest_coverage                REAL,
    asset_turnover                   REAL,
    free_cash_flow_cr                REAL,
    capex_cr                         REAL,
    earnings_per_share               REAL,
    book_value_per_share             REAL,
    dividend_payout_ratio_pct        REAL,
    total_debt_cr                    REAL,
    cash_from_operations_cr          REAL,
    PRIMARY KEY (company_id, year),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);
"""
,
"""
CREATE TABLE peer_groups (
    peer_group_name  TEXT NOT NULL,
    company_id        TEXT NOT NULL,
    benchmark         TEXT,
    PRIMARY KEY (peer_group_name, company_id),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);
"""
,
"""CREATE TABLE market_cap (
    company_id             TEXT NOT NULL,
    year                    INTEGER NOT NULL,
    market_cap_crore         REAL,
    enterprise_value_crore   REAL,
    pe_ratio                 REAL,
    pb_ratio                 REAL,
    ev_ebitda                REAL,
    dividend_yield_pct       REAL,
    PRIMARY KEY (company_id, year),
    FOREIGN KEY (company_id) REFERENCES companies(id)
);
"""
}

7 core + 5 supplementary · load order · load_audit.csv · row counts: companies=92, P&L~1276, BS~1312, CF~1187, prices=5520 · FK check 0

Data Quality Manual Review
5 random companies · year coverage · companies with <5 years · fix loader bugs · re-run

exploratory_queries.sql · unit tests 0 failures · demo nifty100.db · retrospective · update board
Deliverables
nifty100.db — 10 tables populated
output/load_audit.csv — per-table row counts & rejections
output/validation_failures.csv — DQ violations with severity
src/etl/loader.py, validator.py, normaliser.py
db/schema.sql — 10-table SQLite schema
tests/etl/ — 35+ unit tests
notebooks/exploratory_queries.sql (10 queries)

In [58]:
import sqlite3

conn = sqlite3.connect(DB_PATH)

# Get all table names
tables = pd.read_sql_query("""
    SELECT name
    FROM sqlite_master
    WHERE type = 'table'
    ORDER BY name;
""", conn)

print("Tables in database:")
print(tables)

print("\n" + "=" * 60)

# Check each table
for table in tables["name"]:
    
    # Row count
    count = conn.execute(
        f'SELECT COUNT(*) FROM "{table}"'
    ).fetchone()[0]

    # Columns
    columns = pd.read_sql_query(
        f'PRAGMA table_info("{table}")',
        conn
    )

    print(f"\n📌 Table: {table}")
    print(f"Rows: {count}")
    print(f"Columns: {len(columns)}")
    print("Column names:")
    print(columns["name"].tolist())

conn.close()

Tables in database:
                name
0           analysis
1       balancesheet
2           cashflow
3          companies
4          documents
5   financial_ratios
6         market_cap
7        peer_groups
8      profitandloss
9        prosandcons
10           sectors
11      stock_prices


📌 Table: analysis
Rows: 60
Columns: 6
Column names:
['id', 'company_id', 'compounded_sales_growth', 'compounded_profit_growth', 'stock_price_cagr', 'roe']

📌 Table: balancesheet
Rows: 3762
Columns: 13
Column names:
['id', 'company_id', 'year', 'equity_capital', 'reserves', 'borrowings', 'other_liabilities', 'total_liabilities', 'fixed_assets', 'cwip', 'investments', 'other_asset', 'total_assets']

📌 Table: cashflow
Rows: 3509
Columns: 7
Column names:
['id', 'company_id', 'year', 'operating_activity', 'investing_activity', 'financing_activity', 'net_cash_flow']

📌 Table: companies
Rows: 267
Columns: 12
Column names:
['id', 'company_logo', 'company_name', 'chart_link', 'about_company', 'website', '